torch/csrc/jit/script/init.cpp

```diff
@@ -174,89 +163,58 @@ struct VISIBILITY_HIDDEN PythonValue : public SugaredValue {
   py::object self;
 };
 
-// by using torch.jit.Const, a user can mark a python value constant
-// we then make that value immutable.
-// once marked constant, we enable additional behavior such as
-// 1. conversion via asValue to a constant Tensor
-// 2. unrolling of for loops
-struct VISIBILITY_HIDDEN ConstantPythonValue : public PythonValue {
-  using PythonValue::PythonValue;
-  virtual Value * asValue(SourceRange loc, Method & m) override {
+struct VISIBILITY_HIDDEN PythonModuleValue : public PythonValue {
+  explicit PythonModuleValue(py::object mod) : PythonValue(mod) {}
 
-    return PythonValue::asValue(loc, m);
+  std::shared_ptr<SugaredValue> attr(SourceRange loc, Method & m, const std::string& field) override {
+      py::object member = getattr(loc, field);
+      return toSugaredValue(member, m, loc);
   }
-  virtual std::vector<std::shared_ptr<SugaredValue>> asTuple(SourceRange loc, Method& m) override {
-    if(!py::isinstance<py::tuple>(self))
-      return PythonValue::asTuple(loc, m);
+ private:
+};
+
+struct VISIBILITY_HIDDEN BuiltinPythonModuleValue : public PythonModuleValue {
+  explicit BuiltinPythonModuleValue(py::object mod) : PythonModuleValue(mod) {}
+  std::shared_ptr<SugaredValue> attr(SourceRange loc, Method & m, const std::string& field) override {
+    // We support calling functions and using type/layout/device constants
+    // on the torch builtin modules
+    py::object member = getattr(loc, field);
+    if (py::isinstance<py::function>(member)) {
+      return std::make_shared<BuiltinFunction>(field, at::nullopt);
+    }
+    return toSugaredValue(member, m, loc, /*is_constant =*/true);
+  }
+};
 
+bool isBuiltinModule(py::object obj) {
+  // XXX: these can't be static, or they will be destructed after the Python interpreter
+  // exits and that generally sounds like a bad idea
+  py::object torch = py::module::import("torch");
+  py::object functional = py::module::import("torch.nn.functional");
+  return obj.is(torch) || obj.is(functional);
+}
+
+struct VISIBILITY_HIDDEN PythonTupleValue : public PythonValue {
+  explicit PythonTupleValue(py::object tup) : PythonValue(tup) {}
+  std::vector<std::shared_ptr<SugaredValue>> asTuple(SourceRange loc, Method& m) override {
     py::tuple tup = self;
     std::vector<std::shared_ptr<SugaredValue>> result;
-    for(size_t i = 0; i < tup.size(); ++i) {
-      result.push_back(create(loc, m, tup[i]));
+    result.reserve(tup.size());
+    for (size_t i = 0; i < tup.size(); ++i) {
+      // TODO is it safe to always assume tuple elements are constant?
```

Comment:
Yes, the tuple must be constant. You can call this ConstantTupleValue to make it clearer.